### PR TITLE
Fix advanced indexing on "huge" Tensors

### LINKED ISSUE #20888: Reindexing a huge tensor to shuffle it results in data loss at the end of the tensor
## 🐛 Bug

Shuffling a huge tensor by reindexing into it results in all of the values past a certain point being set to 0 when the reindexing is done in GPU memory.

@colesbury  This likely relates to https://github.com/pytorch/pytorch/issues/20562 as the issue is also with reindexing and also shows up at the same point in the tensor.  Hopefully this helps with narrowing the other issue.

I initially thought this was limited to longtensors, but if you use a floattensor with 2x values you run into the exact same issue in the exact same place. 

## To Reproduce

1. Create a huge longtensor (>536870700 values) of random values.  I went with 15Mx45
2. Load it into gpu memory
3. Reindex to shuffle the values

Data past ~536870700 will be all 0.  The data is zero'd mid tensor.

Here's a notebook that reproduces the issue:
https://github.com/EvenOldridge/HugeTensor/blob/master/Huge%20Tensor%20Bug%20-%20Data%20Loss.ipynb

## Expected behavior

Data should not be lost when reindexing

## Environment

- PyTorch Version (e.g., 1.0): 1.01
- OS (e.g., Linux): Linux (Ubuntu 16.04)
- How you installed PyTorch (conda, pip, source): pip
- Python version: 3.6
- CUDA/cuDNN version: 10.0
- GPU models and configuration: TeslaV100 32 Gig (on a DGX-1)

### PR DESCRIPTION
**Pull request by colesbury**
This fixes advanced indexing in cases where there's more than 2^31-1
bytes in the output. The `gpu_index_kernel` was missing the
`can_use_32bit_indexing`/`with_32bit_indexing` check.

This also adds a number of TORCH_INTERNAL_ASSERTS in Loops.cuh,
OffsetCalculator, and IntDivider that sizes are fit in a signed 32-bit
integer.

More comprehensive tests that require a 32 GB GPU are here:
https://gist.github.com/colesbury/e29387f5851521256dff562be07b981e

Fixes #20888

### CODE REVIEW

File: test/test_cuda.py [hunk 1]
```diff
@@ -2456,6 +2456,14 @@ def test(x, ia, ib):
     def test_advancedindex_big(self):
         _TestTorchMixin._test_advancedindex_big(self, lambda t: t.cuda())
 
+    @unittest.skipIf(not TEST_LARGE_TENSOR, "not enough memory")
```
Comment (gchanan) [inline test/test_cuda.py]: should this be a slowTest?  How long does it take?

**Review by gchanan — APPROVED** [review]

File: test/test_cuda.py (hunk 1 above)
Comment (colesbury) [reply test/test_cuda.py]: Yes, it takes ~13 seconds

**Comment by ezyang** [comment]
Looks like this durably broke ROCm https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-test/22251//console (or you were unlucky and an upgrade happened at the same time)

```
04:17:28 test_multinomial (__main__.TestCuda) ... ### HCC STATUS_CHECK Error: HSA_STATUS_ERROR_EXCEPTION (0x1016) at file:mcwamp_hsa.cpp line:1193
04:17:42 Traceback (most recent call last):
04:17:42   File "test/run_test.py", line 441, in <module>
04:17:42     main()
04:17:42   File "test/run_test.py", line 433, in main
04:17:42     raise RuntimeError(message)
04:17:42 RuntimeError: test_cuda failed! Received signal: SIGIOT
```

**Comment by ezyang** [comment]
cc @bddppq @iotamudelta

**Comment by colesbury** [comment]
I think the test failure is unrelated to this change

**Comment by colesbury** [comment]
Well... maybe not. I'll see if https://github.com/pytorch/pytorch/pull/20948 fixes the ROCm build

**Comment by bddppq** [comment]
There was no rocm upgrade recently. Although I don't see anything obviously suspicious in this diff either :(.
Since this has broken master, let's revert it. I will work with @colesbury to locally repro and further debug it tomorrow.

**Pull request closed without merging**